### PR TITLE
Add config file output to --configtest

### LINKED ISSUE #3243: allow outputting resulting config file when merging multiple ones
```
./bin/logstash --show-config -f /my_config_files_dir 
//don't start just show resulting config
input{
//all merged inputs
}
filter{
//all merged filters
}
output{
//all merged outputs
}
```

### PR DESCRIPTION
**Pull request by untergeek**
This only works when the `--debug` flag is also enabled.

fixes #3243

### CODE REVIEW

**Comment by untergeek** [comment]
For reference, this creates output like this: https://gist.github.com/untergeek/e31bedbe5b34f21d1cba

File: lib/logstash/agent.rb [hunk 1]
```diff
@@ -321,10 +321,20 @@ def local_config(path)
         encoding_issue_files << file
       end
       config << cfg + "\n"
+      if config_test?
+        @logger.debug? && @logger.debug("\n***** Begin configuration file: #{file}")
```
Comment (jordansissel) [inline lib/logstash/agent.rb]: Why the newlines?

Also, any thoughts on making this look more like a valid config, so in this case using a comment `#` symbol to prefix the '***** Begin con....' lines?

Comment (jordansissel) [inline lib/logstash/agent.rb]: Let's use structured logging here, too.

`@logger.debug("whatever...", :file => file)`

Comment (untergeek) [inline lib/logstash/agent.rb]: The newlines make it more readable in debug log output, IMO.

I can definitely add the `#`, that's a great idea.  Allows cut/paste testing (but that was the other reason for the newlines).

**Comment by untergeek** [comment]
> Let's use structured logging here, too.
>
>`@logger.debug("whatever...", :file => file)`

I tried that, but it did not output the file readably.  That's the other reason this _only_ outputs like this for `--configtest` _and_ `--debug` output.  People aren't likely to log this to Logstash.  Just stdout or file, for sending to support or a gist.

**Comment by untergeek** [comment]
@jordansissel This is what I get when I try structured logging:

```
Whatever... {:config=>"input {\n  tcp {\n    port => 9500\n  }\n}\n\ninput {\n  tcp {\n    port => 9601\n  }\n}\n\ninput {\n  tcp {\n    port => 9702\n  }\n}\n\ninput {\n  tcp {\n    port => 9399\n  }\n} \n\nfilter {\n  mutate {\n    add_field => { \"a\" => \"b\" }\n  }\n}\n\n\ninput {\n  tcp {\n    port => 9288\n  }\n} \n\nfilter {\n  mutate {\n    add_field => { \"c\" => \"d\" }\n  }\n}\n\n\noutput {\n  stdout { }\n}\n\n", :level=>:debug, :file=>"logstash/agent.rb", :line=>"338", :method=>"local_config"}
```

That's why I went the other way.

**Comment by jordansissel** [comment]
@untergeek I meant specifically for the header-style lines, where you have roughly "The following is the content of a file named #{file}" - my recmomendation is `"The following is the content of a file", :file => file`

**Comment by jordansissel** [comment]
For the actual config itself, I am OK if that appears in the text, like, this line that you have is OK by me:

```diff
+        @logger.debug? && @logger.debug("\n" + cfg + "\n\n")
```

**Comment by untergeek** [comment]
Ah, that.  Yeah, I can address that.

**Comment by jordansissel** [comment]
Otherwise LGTM. Feel free to merge after the `:file => file` change is done. <3

**Comment by untergeek** [comment]
Interesting.  It appears another code change needs to happen a few lines up in 315:

```ruby
@logger.debug("NOT reading config file because it is a temp file", :file => file)
```

I cannot use `:file` as my symbol, because that gets overridden by `logstash/agent.rb` every time.  I'm using `:config_file` instead.

**Comment by jordansissel** [comment]
@untergeek haha oops, yes. Good catch.

**Pull request closed without merging**